Sources/_StringProcessing/Engine/MEBuiltins.swift

```diff
@@ -1,13 +1,221 @@
+  ) -> Bool {
+    guard let c = loadScalar() else {
+      signalFailure()
+      return false
+    }
+
+    var matched: Bool
+    var next = input.unicodeScalars.index(after: currentPosition)
+    switch cc {
+    case .any: matched = true
+    case .anyScalar: matched = true
+    case .anyGrapheme:
+      matched = true
+      next = input.index(after: currentPosition)
+    case .digit:
+      matched = c.properties.numericType != nil && (c.isASCII || !isStrictAscii)
+    case .horizontalWhitespace:
+      matched = c.isHorizontalWhitespace && (c.isASCII || !isStrictAscii)
+    case .verticalWhitespace:
+      matched = c.isNewline && (c.isASCII || !isStrictAscii)
+    case .newlineSequence:
+      matched = c.isNewline && (c.isASCII || !isStrictAscii)
+      if c == "\r" && next != input.endIndex && input.unicodeScalars[next] == "\n" {
+        input.unicodeScalars.formIndex(after: &next)
+      }
+    case .whitespace:
+      matched = c.properties.isWhitespace && (c.isASCII || !isStrictAscii)
+    case .word:
+      matched = (c.properties.isAlphabetic || c == "_") && (c.isASCII || !isStrictAscii)
+    }
+    if isInverted {
+      matched.toggle()
+    }
+    if matched {
+      currentPosition = next
+      return true
+    } else {
+      signalFailure()
+      return false
+    }
+  }
+  
+  func isAtStartOfLine(_ payload: AssertionPayload) -> Bool {
+    if currentPosition == subjectBounds.lowerBound { return true }
+    switch payload.semanticLevel {
+    case .graphemeCluster:
+      return input[input.index(before: currentPosition)].isNewline
+    case .unicodeScalar:
+      return input.unicodeScalars[input.unicodeScalars.index(before: currentPosition)].isNewline
+    }
+  }
+  
+  func isAtEndOfLine(_ payload: AssertionPayload) -> Bool {
+    if currentPosition == subjectBounds.upperBound { return true }
+    switch payload.semanticLevel {
+    case .graphemeCluster:
+      return input[currentPosition].isNewline
+    case .unicodeScalar:
+      return input.unicodeScalars[currentPosition].isNewline
+    }
+  }
+
+  mutating func builtinAssert(by payload: AssertionPayload) throws -> Bool {
+    // Future work: Optimize layout and dispatch
+    switch payload.kind {
+    case .startOfSubject: return currentPosition == subjectBounds.lowerBound
+
+    case .endOfSubjectBeforeNewline:
+      if currentPosition == subjectBounds.upperBound { return true }
+      switch payload.semanticLevel {
+      case .graphemeCluster:
+        return input.index(after: currentPosition) == subjectBounds.upperBound
+         && input[currentPosition].isNewline
+      case .unicodeScalar:
+        return input.unicodeScalars.index(after: currentPosition) == subjectBounds.upperBound
+         && input.unicodeScalars[currentPosition].isNewline
+      }
+
+    case .endOfSubject: return currentPosition == subjectBounds.upperBound
+
+    case .resetStartOfMatch:
+      fatalError("Unreachable, we should have thrown an error during compilation")
```

Comment:
Future: we'll want to version support so that we can reason about back deployment